hydra-plutus/src/Hydra/Contract/OnChain.hs

```diff
@@ -136,20 +133,27 @@ hydraValidatorHash = Scripts.validatorHash . hydraTypedValidator
 
 data Initial
 instance Scripts.ValidatorTypes Initial where
-  type DatumType Initial = PubKeyHash
+  -- NOTE: Ideally, we would like to get rid of validator hashes in here,
+  -- because they are statically known!
+  -- So instead of passing them around, we could simply call
+  -- `hydraValidatorHash` and `commitValidatorHash` in the `initialValidator`...
+  --
+  -- However, this makes the compiler chokes with:
+  --
+  --     GHC Core to PLC plugin: E042:Error: Unsupported feature: Type constructor: GHC.Prim.ByteArray#
+  --
+  -- Hence their presence here.
+  type DatumType Initial = (HeadParameters, ValidatorHash, ValidatorHash, PubKeyHash)
```

Comment:
Not happy with that at moment. Michael suggested to _inject_ the hash in the validator code using template Haskell

(https://input-output-rnd.slack.com/archives/C21UF2WVC/p1627652836016100)

I am a bit afraid of template-haskell stage restriction here but I may look into it later next week.

Reply:
![image](https://user-images.githubusercontent.com/5680256/127667696-799d82c5-7b74-4ce4-8eec-6a6a89ea9764.png)